### ISSUE TITLE
error in create_methyl_summary()

### ISSUE THREAD

**Issue by katieskinner98** [SoundAg/sounDMR#63]
Hi there!

I have some methylation data that I generated using the Oxford Nanopore P2 Solo. I currently have 2 samples with 1 biorep each (human cancer cell line, non-resistant parental line and one drug resistant line). This is a pilot run, hence only 1 biorep. I want to get the pipeline down before I commit to doing more sequencing. 

Methylation calling was done using Dorado. I processed the methyl bam files to get bed methyl files using modkit pileup. I now want to do DMR analysis.

I processed my bed methyl files using the command listed in the readme: 

`${unconverted_bed} | awk -v OFS='\t' '{print $1, $2, $3, $4, $5, $6, $7, $8 ,$9, $10, $11, $4 }' |  sed 's/m,//g' |  sed 's/,0//g' > ${output_bed}`

After this, I used grep in the terminal to subset by chromosome (no differences in bed files compared to running the `split_by_chromosome()` function as far as I can tell, it's just faster). Then as a test I took the first 1000 lines of each chr1 bed file as it was taking too long to run iterations with the full chr1 bed files.

I have successfully managed to create a DMR object, but upon running this DMR object with the `create_methyl_summary()` command I get the error:

`"Output is in a different order. Try running again."`

I also ran my code with the example bed methyl files provided in this repository, and I did not get this error. I am unsure what the issue could be and what to try next. I have attached my files and code. I am wondering if anyone can help me figure out what the issue is?

My R session info is:

```
R version 4.3.3 (2024-02-29)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Sonoma 14.5

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/New_York
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] sounDMR_2.3.2     lubridate_1.9.3   forcats_1.0.0     dplyr_1.1.4       purrr_1.0.2       readr_2.1.5       tidyverse_2.0.0   data.table_1.15.4
 [9] tidyr_1.3.1       tibble_3.2.1      stringr_1.5.1     reshape2_1.4.4    lme4_1.1-35.4     Matrix_1.6-5      glmmTMB_1.1.9     Formula_1.2-5    
[17] ggplot2_3.5.1     changepoint_2.2.4 zoo_1.8-12       

loaded via a namespace (and not attached):
 [1] utf8_1.2.4          generics_0.1.3      stringi_1.8.4       lattice_0.22-6      hms_1.1.3           magrittr_2.0.3      timechange_0.3.0   
 [8] grid_4.3.3          plyr_1.8.9          mgcv_1.9-1          fansi_1.0.6         scales_1.3.0        numDeriv_2016.8-1.1 cli_3.6.3          
[15] rlang_1.1.4         munsell_0.5.1       splines_4.3.3       withr_3.0.0         tools_4.3.3         tzdb_0.4.0          nloptr_2.1.1       
[22] minqa_1.2.7         colorspace_2.1-0    boot_1.3-30         vctrs_0.6.5         R6_2.5.1            lifecycle_1.0.4     MASS_7.3-60.0.1    
[29] pkgconfig_2.0.3     pillar_1.9.0        gtable_0.3.5        glue_1.7.0          Rcpp_1.0.12         tidyselect_1.2.1    rstudioapi_0.16.0  
[36] farver_2.1.2        nlme_3.1-165        labeling_0.4.3      TMB_1.9.13          compiler_4.3.3 
```
Please let me know if additional information is needed. 

Thank you so much in advance for your time! 

Script and files here:
[ks_data.zip](https://github.com/user-attachments/files/16074723/ks_data.zip)

**Comment by decibel-tom** [comment]
Hello! Thank you for your feedback! Looking into your data it seems like you ran Dorado using the 5hmC model, our pipeline is really designed for methylation calls with the 5mC model. One thing you could try is subsetting the bed files to make sure they only have one row per position, so you could just take any rows with an "m" in the 4th or last column. If you do run the latter I'd be curious to hear if that solves the issue!

**Comment by katieskinner98** [comment]
Thanks so much for your response! I tried to subset the bed files to contain only "m", however I am still receiving the same error. I have attached the subsetted files here so you can take a look. 

I was doing some poking around in the data frame resulting from running `generate_methylframe()` and found something odd. Certain positions are duplicated twice, and this appears to be common with those positions with data for both the plus and minus strands. For example:

S1 bed:
chr1 10184 + 100.0 m 4 0
chr1 10184 - 100.0 m 3 0

S2 bed:
chr1 10184 + 100.0 m 5 0
chr1 10184 - 85.71 m 6 1

From my understanding, the resulting data frame for this position should look like this (two rows corresponding to + and - strand):
chr1 10184 + 100.0 m 4 0 + 100.0 5 0 0 
chr1 10184 - 100.0 m 3 0 - 85.71 m 6 1

However, what I am actually getting is this (4 rows, 2 correct, and 2 incorrect):
chr1 10184 + m 100.0 4 0 100.0 5 0 [seems correct]
chr1 10184 + m 100.0 4 0 85.71 6 1 [incorrect, is mixing S1 + strand with S2 - strand]
chr1 10184 - m 100.0 3 0 100.0 5 0 [incorrect, mixing S1 - strand with S2 + strand]
chr1 10184 - m 100.0 3 0 85.71 6 1 [seems correct]

When I try to rerun the analysis by subsetting with just the + strand (files also attached), I do not get the same error, therefore leading me to believe that there is something odd happening in the generation of the methylframe that could have something to do with the strand information. 

Thanks so much for your help! Let me know what you think!

[subsetdata.zip](https://github.com/user-attachments/files/16167750/subsetdata.zip)

**Comment by jcolicchio-soundag** [comment]
Hey Katie, thanks for the info!  Our pipleline is using the assumption that 5mC methylation can only be on either the + or - strand, but not both.  For this reason, it assumes only one row per position. Using the standard 5mC all context model from dorado, this should always be the case.  You are correct that having both a + and - strand methylation call at the same position is responsible for this issue.  

I am not quite sure what the biological situation is where the model you used is assigning methylation to both + and - strand, unless it has to do with genetic variation with a (C/G) snp, such that either steand can be methylated.  As far as options go, I would be inclined to re-run Dorado using the 5mC model, but if these results do look biologically realistic then there are other possible workarounds. One option would be to merge the +/- columns into a single row beforehand, or filter out these rows. 

Best,
Jack

**Comment by katieskinner98** [comment]
Hi Jack,

Thanks so much for your response! 

As mentioned in my original post, methylation calling was done using Dorado. The only modified base calling model I was able to choose from was "none" or "5mC & 5hmC CG contexts" (this is done using the MinKNOW UI). 

My downstream processing steps from there were: merging modified bams, aligning with minimap2, then sorting and indexing the merged/aligned bam files. From there I used modkit pileup and then the processing steps to get the methyl bed file into the format required for your package.

Thanks again for your response. I'll try playing around with my data and see if I can get something to work.

Katie
